### ISSUE TITLE
Sequelize + Webpack in 2021 - Can you help me?

### ISSUE THREAD

**Issue by papb** [sequelize/sequelize#13169]
Hello everyone. Many people have reported problems with Sequelize + Webpack in the last few years. I can guess that the root of the problem is some kind of dynamic require involving the optional dialects supported by Sequelize.

However, I have had a hard time exploring the problem for a few reasons:

* I never used Webpack
* Most resources I found on the internet were thinking of Webpack for the frontend (which is probably the most popular usage)
* The resources I found about Webpack for backend seemed very outdated
* My quick attempts of reproducing the problem failed
* I also haven't used things like Serverless, Azure and AWS Lambda (which seem to be mentioned in the same issues) - I don't have enough knowledge to decide whether or not those are just [red herrings](https://en.wikipedia.org/wiki/Red_herring) and all issues stem from the first cause or not.
* Most issues were opened several years ago and I couldn't be sure if the original problem (mentioned in the first post) still happens or if people are just using the related issue to comment about related (but not-exactly-the-same) problems.
* My lack of experience/knowledge with bundlers like Webpack does not allow me to evaluate if the problems are Sequelize's fault, Webpack's fault or the users' fault (like a misconfiguration of some kind).

I noticed two main types of issues:

* **\["Type A"\]** Errors when executing Webpack, usually in the form of `Can't resolve {some-package}`, such as this one from 2017: #7509
* **\["Type B"\]** Errors when running code processed by Webpack, usually in the form of `Please install {some-package} manually`, such as this one from 2018: #9489

My first guess is that the problem is in [this code](https://github.com/sequelize/sequelize/blob/9f950cbcbdd659d559496b77c40e0f827b108561/lib/dialects/abstract/connection-manager.js#L66). Especially since the other place which used to perform dynamic requires in our source code [was refactored in 2019](https://github.com/sequelize/sequelize/pull/11464/files#diff-fd47c0d1a3753dc2af16c6d8761d18ebe50f4d4a44d1cffa87fb8f4332277d92) via #11464.

However, I need to really understand the problem in order to fix it, because I want to write a failing test first, and only then fix this. The failing test will probably be a separate workflow for GitHub actions which is going to really install Webpack and run it and see that it works.

**Can anyone create a new [MCVE](https://stackoverflow.com/help/minimal-reproducible-example) for both cases (Type A and Type B) in the form of a public GitHub repository that I can clone and run `npm install && npm start` to see the problem?**

I will then use that as a starting point for the new GitHub workflow that is going to test this.

**Comment by papb** [comment]
By the way, here is my quick (failed) attempt to see the first type of problem happening myself (after following a quick Webpack setup from a tutorial):

```js
// Entrypoint
import { Sequelize } from 'sequelize';
```

```js
// webpack.config.js
const fs = require('fs');
module.exports = {
  // ...
  target: 'node',
  externals: Object.fromEntries(
    fs.readdirSync('node_modules').filter(x => x !== 'bin').map(x => [x, 'commonjs ' + x])
  )
};
```

**Comment by papb** [comment]
After looking at #9608, I am wondering if the `dialectModule` option solves at least part of these problems?

**Comment by AwolDes** [comment]
@papb from what I know & from what I have experienced, you are right when you say: 
> the root of the problem is some kind of dynamic require involving the optional dialects supported by Sequelize.

Webpack is used to bundle javascript, both on the frontend & on the server (especially when using Lambda and similar serverless solutions).

I fixed both "Type A" and "Type B" errors by using `require` syntax, and not `import` syntax for packages, particularly Sequelize (although I think this might come down to my configuration, I'm not 100% sure). My PoC repo https://github.com/AwolDes/express-sequlize-lambda-webpack-poc mentioned in https://github.com/sequelize/sequelize/issues/12874 might be a good place to start to try and re-produce. I wonder if adding the changes in https://github.com/sequelize/sequelize/pull/9609 would help

**Comment by Trinovantes** [comment]
Here's an example for type A using latest Webpack https://github.com/Trinovantes/issue-sequelize-webpack

As I mentioned this in https://github.com/sequelize/sequelize/issues/7509#issuecomment-754345526, Webpack bundles every dependency into a single file. If it can't know for sure at compile time that something may/may not be needed, it includes it to be safe:

https://github.com/sequelize/sequelize/blob/a8b8b7f3c439f0b48b6ccc65ba4953b13aeabe86/src/data-types.js#L1023-L1027

From Webpack's perspective, all of these `require`s *might* be executed at runtime so it'll try to bundle them and fail

**Comment by Keimeno** [comment]
A possible solution might be to dynamically require all dialects:

```js
const dialects = ['postgres', 'mysql', ...];
dialects.forEach(dialect => dialectMap[dialect] = require(`./dialects/${dialect}/data-types`)(DataTypes));
```

or only require if a given dialect should be used

```js
// input from developer
const dialect = 'postgres';
dialectMap[dialect] = require(`./dialects/${dialect}/data-types`)(DataTypes);
```

But this could potentially still cause problems with webpack, will investigate this further.

**Comment by Trinovantes** [comment]
I think a plugin system would work:
```ts
import { Sequelize } from 'sequlize'
import MysqlDialect from 'sequelize/plugins/mysql'
Sequelize.use(MysqlDialect)
```

This is similar to how other libraries like dayjs handle optional components and compatibility with webpack https://day.js.org/docs/en/plugin/relative-time

**Comment by HirenPatel2791** [comment]
This worked for me for [serverless-bundle](https://www.npmjs.com/package/serverless-bundle).

```
"serverless": "^2.10.0",
"node": "12.x"
"serverless-bundle": "^3.2.1"
"sequelize": "^6.6.2"
```
**serverless.yml** (this solves: 'Module not found: Error: Can't resolve 'pg-hstore' in '/Users/.../node_modules/sequelize/lib/dialects/postgres')
```
custom:
  bundle:
      ignorePackages: 
        - pg-hstore
```
Now since it's ignoring the pg dialect module, I need to define my dialect module like when init sequelize.
```
const { Sequelize } = require('sequelize');
const mysql = require('mysql2');

const sequelize = new Sequelize(
    process.env.DB_NAME || 'blah',
    process.env.DB_USER || 'blah',
    process.env.DB_PASS || 'blah',
    {
        host: process.env.DB_URL || 'localhost',
        dialect: 'mysql',
        dialectModule: mysql
    }
);

export default sequelize;
```

**Comment by invaderb** [comment]
I was able to get sequelize working with serverless-bundle with @HirenPatel2791 recommendation, but not with just serverless-webpack.

serverless bundle doesn't leave much for configuration and no minification so the files are much larger.

@papb with serverless-webpack I tried the dialectModule and was still getting errors and warnings.

**Comment by jkpet** [comment]
January 2023 and this is still an issue (for me at least). Haven't been able to bundle it successfully despite the suggestions above (using sqlite)

**Comment by WikiRik** [comment]
We can reopen this issue when we have more information to work with. I want to especially highlight this part of the original post;

> However, I need to really understand the problem in order to fix it, because I want to write a failing test first, and only then fix this. The failing test will probably be a separate workflow for GitHub actions which is going to really install Webpack and run it and see that it works.
> 
> **Can anyone create a new [MCVE](https://stackoverflow.com/help/minimal-reproducible-example) for both cases (Type A and Type B) in the form of a public GitHub repository that I can clone and run `npm install && npm start` to see the problem?**
> 
> I will then use that as a starting point for the new GitHub workflow that is going to test this.

**Comment by bhallstein** [comment]
@willrpike Here is a repo which minimally replicates the issue: https://github.com/bhallstein/sequelize-bundling-issue

This repo uses rollup, which emits a warning during build (`Unresolved dependencies... pg-hstore (imported by "pg-hstore?commonjs-external")`). Then when running the built bundle, node produces the hard error when it attempts to import the nonexistent dependency. The underlay issue is almost certainly identical when using webpack, implementing a fix for one should also resolve this for the other.

The Readme contains instructions for replicating but it's as simple as:

```sh
npm install
npm run build
node dist/app.js
```

I hope that's of some help.

**Comment by ephys** [comment]
You need to install `pg-hstore` if you want to bundle sequelize instead of loading it as an external

In Sequelize 6, `pg-hstore` is an optional dependency that you must install if you use `DataTypes.HSTORE`. It's [loaded dynamically](https://github.com/sequelize/sequelize/blob/fd4afa6a89c111c6d6d0c94f0b98bf421b5357b6/src/dialects/postgres/data-types.js#L357)

Rollup can't do that so it tries to bundle it, so you need to either make it an external or install it

In Sequelize 7, `pg-hstore` is [always installed](https://github.com/sequelize/sequelize/blob/7713cb2d6a16dffd885bcfbb1e2a80913fb74490/packages/core/src/dialects/postgres/data-types.ts#L13) because it's a tiny library and we're going to move each dialect to its own package (e.g. `@sequelize/postgres`) which should resolve this problem.

**Comment by bhallstein** [comment]
Thanks very much for the informative reply @ephys! Simply installing pg-hstore does seem to simply resolve the missing dependency issue, and as you say, it's not a large dependency.

I looked further into config on this rollup-side, and the commonjs plugin can handle dynamic imports can be handled with config similar to this:

```js
commonjs({
  include: /node_modules/,
  dynamicRequireTargets: [
    '**/node_modules/sequelize/**/*.js',
  ],
  dynamicRequireRoot: `node_modules`,
  transformMixedEsModules: true,
}),
```

(Currently blocked on fully testing this due to another tooling issue, will update this comment when able to test fully.)

The dialect split sounds like a good resolution for the future.

Thanks for your help.

**Comment by Zenlight** [comment]
Hello all! I happen to just receive the same error as above and one more. The first one went away when I installed `pg-hstore`.
However, now I'm left with this error:
```
vue    |  ERROR  Failed to compile with 1 error6:14:41 PM
vue    | 
vue    | This dependency was not found:
vue    | 
vue    | * node:util in /node_modules/sequelize/lib/dialects/db2/query.js
vue    | 
vue    | To install it, you can run: npm install --save node:util
vue    | <s> [webpack.Progress] 100% 
```

Ive tried the` externals` and `plugins` fix, but I'm still receiving the error above. 
Any help is appreciated, thank you!

**Comment by bhallstein** [comment]
@Zenlight Vue builds for the browser, where node’s library functions aren’t available – see https://github.com/sequelize/sequelize/issues/6424

You might be able to shim the node library using browserify, but you’re likely to be better off using a library built for the browser. Maybe have a look at knex?

**Comment by Zenlight** [comment]
Hi @bhallstein. I will look into that tomorrow and get back you. Thanks for the quick response!

**Comment by xlinxie** [comment]
> After looking at #9608, I am wondering if the `dialectModule` option solves at least part of these problems?

`dialectModule` solves Please install {some-package} manually problem

**Comment by sallyzoughaib** [comment]
@bhallstein thank for you,
I am attempting to bundle my Node.js TypeScript backend application code and dependencies (node_modules) into a single JavaScript file named "bundle.js".

I got this error when i ran npm run build

where build : rollup -c

Error: [!] (plugin rpt2) Error: Unexpected character '�' (Note that you need plugins to import files that are not JavaScript)node_modules/@sentry/profiling-node/lib/sentry_cpu_profiler-darwin-x64-93.node (1:0)

This is my rollup.config.js filer:

import alias from '@rollup/plugin-alias';
import nodeBuiltins from 'rollup-plugin-node-builtins';
import commonjs from 'rollup-plugin-commonjs';
import resolve from 'rollup-plugin-node-resolve';
import typescript from 'rollup-plugin-typescript2';
import tslibResolveId from 'rollup-plugin-tslib-resolve-id';
import json from '@rollup/plugin-json';

export default {
  input: 'src/server.ts',
  output: {
    file: 'dist/bundle.js',
    format: 'es',
  },
  plugins: [
    nodeBuiltins(),
    tslibResolveId(),
    typescript(),
    json(),
    commonjs({
      include: 'node_modules/**',
      ignoreGlobal: false,
      sourceMap: false,
    }),
    alias({
      entries: [
        { find: '@config', replacement: 'config' },
        { find: '@controllers', replacement: 'controllers' },
        { find: '@middlewares', replacement: 'middlewares' },
        { find: '@routes', replacement: 'routes' },
        { find: '@services', replacement: 'services' },
      ],
    }),
    resolve(),
  ],
};

**Comment by ephys** [comment]
All dynamic requires are being phased out in Sequelize 7. `_loadDialectModule` has already been fully removed. These changes are not released yet and are tracked in #13722

This should make Sequelize itself bundable, but I can't speak for the libraries we use that handle connecting to the database

**Comment by ephys** [comment]
Unfortunately that is not a change we can implement as it would require all users to install `mysql2`, instead of just the users that are using the mysql dialect

If you use `dialectModule`, try setting it the result of `import * as mysql2 from "mysql2";` instead of `import mysql2 from "mysql2";`

**Comment by cirosantilli** [comment]
The following is a minimal example of "Please install {some-package} manually". If you replace sequelize with webpack compatible package like lodash it works.

Adding:

```
    "pg-hstore": "^2.3.4",
```

as a workaround to another issue even though it should not be needed in this sqlite example.

webpack.config.js

```
const { ProvidePlugin } = require("webpack")

module.exports = {
  entry: {
    index: ['./index.js'],
  },
  mode: 'none',
  output: {
    filename: '[name].js',
  },
  target: 'node',

  // Blows up differently.
  //externals: {
  //  sequelize: 'sequelize',
  //},

  // Did not help.
  //plugins: [
  //  new ProvidePlugin({
  //    sqlite3: 'sqlite3',
  //  })
  //]
}
```

package.json

```
{
  "name": "webpack-cheat",
  "engines": {
    "node": ">=v20.10.0"
  },
  "version": "1.0.0",
  "private": true,
  "scripts": {
    "build": "webpack"
  },
  "author": "Ciro Santilli",
  "license": "MIT",
  "devDependencies": {
    "webpack": "5.93.0",
    "webpack-cli": "5.1.4",
    "webpack-dev-server": "5.0.4"
  },
  "dependencies": {
    "pg-hstore": "^2.3.4",
    "sequelize": "6.14.0",
    "sqlite3": "5.0.2"
  }
}
```


index.js

```
const assert = require('assert')
const path = require('path')

/*
Might move things forward, but fails instead with an error that likely comes from sqlite3
also not being compatible with webpack:

ERROR in ./node_modules/node-pre-gyp/lib/info.js 14:14-32
Module not found: Error: Can't resolve 'aws-sdk' in '/home/ciro/bak/git/cirosantilli.github.io/webpack/sequelize/node_modules/node-pre-gyp/lib'
 @ ./node_modules/node-pre-gyp/lib/ sync ^\.\/.*$ ./info.js ./info
 @ ./node_modules/node-pre-gyp/lib/node-pre-gyp.js 52:13-36 184:38-55
 @ ./node_modules/sqlite3/lib/sqlite3-binding.js 1:13-36
 @ ./node_modules/sqlite3/lib/sqlite3.js 2:14-45
 @ ./index.js 3:16-34
*/
//const sqlite3 = require('sqlite3')
const { DataTypes, Sequelize } = require('sequelize')
let sequelize
if (process.argv[2] === 'p') {
  sequelize = new Sequelize('tmp', undefined, undefined, {
    dialect: 'postgres',
    host: '/var/run/postgresql',
  })
} else {
  sequelize = new Sequelize({
    dialect: 'sqlite',
    storage: 'tmp.sqlite',
  })
}
function assertEqual(rows, rowsExpect) {
  assert.strictEqual(rows.length, rowsExpect.length)
  for (let i = 0; i < rows.length; i++) {
    let row = rows[i]
    let rowExpect = rowsExpect[i]
    for (let key in rowExpect) {
      assert.strictEqual(row[key], rowExpect[key])
    }
  }
}
;(async () => {
const IntegerNames = sequelize.define('IntegerNames', {
  value: { type: DataTypes.INTEGER },
  name: { type: DataTypes.STRING },
});
await IntegerNames.sync({ force: true })
async function reset() {
  await sequelize.truncate({ cascade: true })
  await IntegerNames.create({ value: 2, name: 'two' })
  await IntegerNames.create({ value: 3, name: 'three' })
  await IntegerNames.create({ value: 5, name: 'five' })
}
await reset()
let rows
rows = await IntegerNames.findAll({ order: [['value', 'ASC']] })
assertEqual(rows, [
  { id: 1, value: 2, name: 'two',   },
  { id: 2, value: 3, name: 'three', },
  { id: 3, value: 5, name: 'five',  },
])
})().finally(() => { return sequelize.close() })
```
